src/components/FromBob/bobConversionUtils.ts

```diff
@@ -0,0 +1,242 @@
+}
+
+interface FunctionSubstitutionInterface {
+  [key: string]: (
+    inputProps: UnknownPropsObject,
+    ouptutProps: UnknownPropsObject
+  ) => void;
+}
+
+export const bobMacrosToMacroMap = (
+  inputProps: UnknownPropsObject,
+  outputProps: UnknownPropsObject
+): void => {
+  // Start with blank object
+  if (inputProps.macros) {
+    outputProps.macroMap = {};
+    Object.entries(inputProps.macros as object).forEach(
+      ([key, value]): void => {
+        outputProps.macroMap[key] = value["_text"];
+      }
+    );
+  }
+};
+
+export interface BobColor {
+  _attributes: { name: string; red: string; blue: string; green: string };
+}
+
+export const bobColorsToColor = (color: BobColor): string => {
+  try {
+    return `rgb(${color._attributes.red}, ${color._attributes.green}, ${color._attributes.blue})`;
+  } catch (e) {
+    log.error(`Could not convert color object`);
+    log.error(color);
+    return "";
+  }
+};
+
+export const bobBackgroundColor = (
+  inputProps: UnknownPropsObject,
+  outputProps: UnknownPropsObject
+): void => {
+  outputProps.backgroundColor = bobColorsToColor(
+    inputProps.background_color.color
+  );
+};
+
+export const bobForegroundColor = (
+  inputProps: UnknownPropsObject,
+  outputProps: UnknownPropsObject
+): void => {
+  outputProps.color = bobColorsToColor(inputProps.foreground_color.color);
+};
+
+export const bobPrecisionToNumber = (
+  inputProps: UnknownPropsObject,
+  outputProps: UnknownPropsObject
+): void => {
+  try {
+    outputProps.precision = Number(inputProps.precision._text);
+  } catch (e) {
+    log.error(
+      `Could not convert precision of ${inputProps.precision} to a number`
+    );
+  }
+};
+
+export const bobVisibleToBoolen = (
+  inputProps: UnknownPropsObject,
+  outputProps: UnknownPropsObject
+): void => {
+  try {
+    let visible = inputProps.visible._text;
+    if (visible === "true") {
+      outputProps.visible = true;
+    } else if (visible === "false") {
+      outputProps.visible = false;
+    }
+  } catch (e) {
+    log.error(
+      `Could not convert visible property ${inputProps.visible} to a number`
+    );
+  }
+};
+
+export const bobAvoidStyleProp = (
+  inputProps: UnknownPropsObject,
+  outputProps: UnknownPropsObject
+): void => {};
+
+export const bobActionToAction = (
+  bobAction: UnknownPropsObject
+): WidgetActions => {
+  let actionsToProcess: any[] = [];
+  if (Array.isArray(bobAction)) {
+    actionsToProcess = bobAction;
+  } else {
+    actionsToProcess = [bobAction];
+  }
+
+  // Object of available actions
+  const availableActions: { [key: string]: any } = {
+    write_pv: WRITE_PV, // eslint-disable-line @typescript-eslint/camelcase
+    WRITE_PV: WRITE_PV
+  };
+
+  // Turn into an array of Actions
+  let processedActions: WidgetActions = { executeAsOne: false, actions: [] };
+
+  actionsToProcess.forEach((action): void => {
+    log.debug(action);
+    let type: string = availableActions[action._attributes.type];
```

Comment:
Add an error handler here for the case where the action type is not found

Otherwise we will try and convert a slightly different opi or bob file and the whole thing will immediately crash

Reply:
Done, just notes the error in the console and moves on as with the other conversion utils